### PR TITLE
[WIP] Add ws_wayland_obj type

### PR DESCRIPTION
**Pull request by matthiasbeyer**
This PR implements the `ws_wayland_obj` type.

Can someone suggest more implementation details, e.g. which kind of
methods should be provided?

### CODE REVIEW

**Comment by neithernut** [comment]
Info: I'm currently evaluating what needs to be added to this type.

**Comment by neithernut** [comment]
We should simply incorporate the `wl_object` into this object type.
Or possibly a `wl_resource`

**Comment by matthiasbeyer** [comment]
Targets #112 .

File: src/objects/wayland_obj.h [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+ */
+extern ws_object_type_id WS_OBJECT_TYPE_ID_WAYLAND_OBJ;
+
+/**
+ * Get a newly allocated and initialized `ws_wayland_obj` object
+ *
+ * @return new `ws_wayland_obj` object or NULL on failure
+ */
+struct ws_wayland_obj*
+ws_wayland_obj_new(void);
```
Comment (neithernut) [inline src/objects/wayland_obj.h]: Could take an `wl_object` as parameter.

**Comment by matthiasbeyer** [comment]
More methods for this type? Suggestions?

**Comment by neithernut** [comment]
Please add methods for:
- getting a `wl_resource`
- getting the serial of an object.

**Comment by matthiasbeyer** [comment]
After searching the wayland docu (which is shit, btw): A serial can only be retreived from a `wl_display` object, which is _not_ a subtype of `wl_object`.

**Comment by matthiasbeyer** [comment]
A `wl_resource` is a subtype of a `wl_object` and can therefor retreived from our `ws_wayland_obj` type with the method I just added.

**Comment by neithernut** [comment]
A `wl_object` _always_ has a serial assigned to it!
The serial is _generated_ by `wl_display`.

**Comment by neithernut** [comment]
Add a hash callback which calculates the hash by dividing `SIZE_MAX` by the object's id.

File: src/objects/wayland_obj.c [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+/**
+ * Type information for ws_wayland_obj type
+ */
+ws_object_type_id WS_OBJECT_TYPE_ID_WAYLAND_OBJ = {
+    .supertype  = &WS_OBJECT_TYPE_ID_OBJECT,
+    .typestr    = "ws_wayland_obj",
+
+    .init_callback = init_callback,
+};
```
Comment (neithernut) [inline src/objects/wayland_obj.c:63]: `NULL` the unused callbacks.

**Comment by neithernut** [comment]
We need `ws_wayland_obj_init()`.

---

Note: this PR is blocking quite a few things, so please give it some priority.
From my part, you're authorized to squash things once you're done, to speed things up.

**Comment by neithernut** [comment]
Btw: I'm still for returning a `wl_resource`. If you have a look at the API, most functions will take a resource. Considering also that casing from a resource to an object can be done by accessign the element in the resource, returning the resource is much cleaner.
Anyways: casting in one spot instead of all over the place is preferable, so please return a `el_resource` and _not_ an `wl_object`.

**Comment by matthiasbeyer** [comment]
What about that, you can either get the object behind `ws_wayland_obj` as `wl_resource` or as `wl_object` now.

I'm not very happy with this, as you don't know what is actually stored in the `ws_wayland_obj` object, ...

File: src/objects/wayland_obj.c [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    if (0 == ws_wayland_obj_init(w, r)) {
+        return w;
+    }
+
+    free(w);
+    return NULL;
+}
+
+struct wl_object*
+ws_wayland_obj_get_wl_obj(
```
Comment (neithernut) [inline src/objects/wayland_obj.c]: You forgot to remove this.

**Comment by neithernut** [comment]
ACK.

**Pull request merged**